File: src/a03.py
trees = set()
for r, line in enumerate(open('input03.txt').readlines()):
    line = line.strip()
    for c, symbol in enumerate(line):
        if symbol == '#':
            trees.add((r, c))
width = len(line)


def count_trees(slope, trees, r, width):
    pos = (0, 0)
    tree_count = 0
    while pos[0] <= r:
        new_pos = pos[0] + slope[0], (pos[1] + slope[1]) % width
        if new_pos in trees:
            tree_count += 1
        pos = new_pos

    return tree_count


print(f"part 1: {count_trees(slope=(1, 3), trees=trees, r=r, width=width)}")


def prod(array):
    if len(array) > 1:
        return array[0] * prod(array[1:])
    else:
        return array[0]


slopes = [(1, 1), (1, 3), (1, 5), (1, 7), (2, 1)]
print(f"part 2: {prod([count_trees(slope, trees, r, width) for slope in slopes])}")

###############################################################################

File: src/a03.ts
import { readFileSync } from "fs";

var lines = readFileSync("input/input03.txt", "utf-8").trim().split("\n");

var trees = new Set();

for (var r = 0; r < lines.length; r++) {
  var line = lines[r];
  for (var c = 0; c < line.length; c++) {
    if (line[c] == "#") {
      trees.add([r, c].toString());
    }
  }
}

function countCollisions(slope: number[]) {
  var pos = [0, 0];

  var collisions = 0;
  while (pos[0] < r) {
    pos[0] += slope[0];
    pos[1] = (pos[1] + slope[1]) % c;
    if (trees.has([pos[0], pos[1]].toString())) {
      collisions += 1;
    }
  }
  return collisions;
}

console.log("part 1:", countCollisions([1, 3]));

// part 2
var part2 =
  countCollisions([1, 1]) *
  countCollisions([1, 3]) *
  countCollisions([1, 5]) *
  countCollisions([1, 7]) *
  countCollisions([2, 1]);

console.log("part2:", part2);

// Things I've learned
// Sets in javascript also have the hashing problem they have in Python
// I had to resort to strings instead of tuples as I would ahve in Python
